### PR TITLE
Adding support for Python 3.9

### LINKED ISSUE #631: Add support for Python 3.9
### Problem Description

SDV should support Python 3.9

### PR DESCRIPTION
**Pull request by amontanez24**
Resolves #631

### CODE REVIEW

File: sdv/tabular/base.py [hunk 1]
```diff
@@ -234,7 +234,6 @@ def _sample_rows(self, num_rows, conditions=None, transformed_conditions=None,
                     sampled = self._sample(num_rows, transformed_conditions)
                 except NotImplementedError:
                     sampled = self._sample(num_rows)
-
```
Comment (pvk-developer) [inline sdv/tabular/base.py:237]: Maybe this was deleted by mistake, but I think it should stay there.

**Review by pvk-developer — CHANGES_REQUESTED** [review]
Looks good and it works as expected.

File: sdv/metadata/table.py [hunk 2]
```diff
@@ -661,7 +661,10 @@ def transform(self, data, on_missing_column='error'):
         data = self._transform_constraints(data, on_missing_column)
 
         LOGGER.debug('Transforming table %s', self.name)
-        return self._hyper_transformer.transform(data)
+        try:
+            return self._hyper_transformer.transform(data)
+        except rdt.errors.NotFittedError:
+            return data
```
Comment (csala) [inline sdv/metadata/table.py:667]: Why do we return `data` if `NotFitted`? Should we not let the error be raised?

File: sdv/metadata/table.py [hunk 3]
```diff
@@ -693,7 +696,10 @@ def reverse_transform(self, data):
         if not self.fitted:
             raise MetadataNotFittedError()
 
-        reversed_data = self._hyper_transformer.reverse_transform(data)
+        try:
+            reversed_data = self._hyper_transformer.reverse_transform(data)
+        except rdt.errors.NotFittedError:
+            reversed_data = data
```
Comment (csala) [inline sdv/metadata/table.py:702]: Same comment as above

File: sdv/constraints/tabular.py [hunk 4]
```diff
@@ -1028,11 +1029,18 @@ def _sample_constraint_columns(self, table_data):
                         proposed_table_data[column] = 1.0
 
             else:
+                should_transform = True
+                conditions = sub_table_data[condition_columns]
+                transformed_conditions = self._hyper_transformer.transform(conditions)
                 proposed_table_data = self._columns_model.sample(
                     num_rows=sub_table_data[condition_columns].shape[0],
```
Comment (csala) [inline sdv/constraints/tabular.py:1036]: Isn't this just `num_rows=len(sub_table_data)`?

File: sdv/constraints/tabular.py [hunk 5]
```diff
@@ -1020,6 +1020,7 @@ def _sample_constraint_columns(self, table_data):
         has_one = table_data[condition_columns].sum(axis=1) == 1.0
```
Comment (csala) [inline sdv/constraints/tabular.py:1020]: Not really part of this PR, but we could think about doing the fix here:
It seems like we are accessing `table_data[condition_columns]` several times, and also computing `sum(axis=1)` on them at least twice. Could we capture these in variables?

File: sdv/constraints/tabular.py [hunk 6]
```diff
@@ -1028,11 +1029,18 @@ def _sample_constraint_columns(self, table_data):
                         proposed_table_data[column] = 1.0
 
             else:
+                should_transform = True
+                conditions = sub_table_data[condition_columns]
+                transformed_conditions = self._hyper_transformer.transform(conditions)
```
Comment (csala) [inline sdv/constraints/tabular.py:1034]: It seems like this introduces a `transform` that was not originally done, so the behavior is being changed. Why is this done?

File: sdv/metadata/table.py (hunk 2 above)
Comment (amontanez24) [reply sdv/metadata/table.py:667]: This happens when data that is all primary keys gets passed. The `HyperTransformer` raises a not fitted error even though it was fit because the `_transformers_sequence` is empty

File: sdv/constraints/tabular.py (hunk 6 above)
Comment (amontanez24) [reply sdv/constraints/tabular.py:1034]: It should have been transformed before. It wasn't because the values are numeric and the column names were the same before so it didn't matter. Now that the column names are different, the model doesn't know the conditions' original names, and they should be transformed.

File: sdv/metadata/table.py (hunk 2 above)
Comment (csala) [reply sdv/metadata/table.py:667]: I see. So it seems to rather be a bug on RDT, right?
I think the change is fine, though, because this code is protected by the `if not self.fitted` above, meaning that this `rdt.errors.NotFitted` will only be raised in that scenario that you described.

Comment (amontanez24) [reply sdv/metadata/table.py:667]: Yeah, I think we will have to fix this in RDT

**Review by csala — APPROVED** [review]
After resolving some doubts, this looks good to go.
@amontanez24 it would be great if we can add the small fixes I commented that are unrelated to the PR, but it's also OK to merge as it is and open an issue to come back and optimize this some other time.

File: sdv/constraints/tabular.py (hunk 4 above)
Comment (amontanez24) [reply sdv/constraints/tabular.py:1036]: Hmm I think so, Idk why this is like that

**Review by pvk-developer — APPROVED** [review]
:shipit:
